hyperspy/misc/array_tools.py

```diff
@@ -96,11 +97,106 @@ def rebin(a, new_shape):
     new_shape = tuple(int(ns) for ns in new_shape)
     factor = np.asarray(shape) // np.asarray(new_shape)
     evList = ['a.reshape('] + \
-             ['new_shape[%d],factor[%d],' % (i, i) for i in range(lenShape)] + \
+             ['new_shape[%d],factor[%d],' % (i, i) for i in range(lenShape)] +\
              [')'] + ['.sum(%d)' % (i + 1) for i in range(lenShape)]
     return eval(''.join(evList))
 
 
+def _linear_bin(s, scale,
```

Comment:
I think that it should be possible to improve the performance by i) rewriting it in C or Cython ii) vectorizing it. iii) using numba? Now that we distribute some cython code, i) doesn't look like a bad idea. I have no experience with numba, but, although it seems to be perfect for the job, it is probably not straightforward to install it in anything but anaconda. However, the easiest approach might be to "vectorize it". I think that it should be possible to write this functions as the standard `rebin` followed by a postprocessing step. For example, say that you want to rebin a list of 5 numbers, `a`, using 2.5 bins. You could do it like this:

```python

rebin_size = floor(scale)
a2rebin = a[0, 1, 3, 4]
result = rebin(a2rebin, ((2,)))
result[0] = (a[1] + a[2]) / 2
result[1] = (a[1] + a[2]) / 2

```

This is of course the simplest example that I've been able to come with, but I think that it should be possible to write it so that it works in the general case by automatically generating the index to obtain `a2rebin` and the interpolation step at the end. What do you think?